### PR TITLE
Migrate TaxCategorySync

### PR DESCRIPTION
**Pull request by salander85**
based on migration of sync-utils.
JIRA: https://commercetools.atlassian.net/browse/DEVX-159

### CODE REVIEW

File: src/integration-test/java/com/commercetools/sync/integration/sdk2/commons/utils/TaxCategoryITUtils.java [hunk 1]
```diff
@@ -87,47 +86,49 @@ private static CompletionStage<TaxCategory> deleteTaxCategoryWithRetry(
...
+                                  .key(TAXCATEGORY_KEY)
+                                  .build();
+                          return createTaxCategory(taxCategoryDraft, ctpClient);
+                        }))
+        .toCompletableFuture()
+        .join();
+  }
 
-    return taxCategory;
+  static CompletionStage<Optional<TaxCategory>> taxCategoryExists(
```
Comment (lojzatran) [inline src/integration-test/java/com/commercetools/sync/integration/sdk2/commons/utils/TaxCategoryITUtils.java:109]:
```suggestion
  private static CompletionStage<Optional<TaxCategory>> taxCategoryExists(
```

File: src/main/java/com/commercetools/sync/sdk2/taxcategories/TaxCategorySync.java [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+
+  private final TaxCategoryService taxCategoryService;
+  private final TaxCategoryBatchValidator batchValidator;
+
+  public TaxCategorySync(@Nonnull final TaxCategorySyncOptions taxCategorySyncOptions) {
+    this(taxCategorySyncOptions, new TaxCategoryServiceImpl(taxCategorySyncOptions));
+  }
+
+  /**
+   * Takes a {@link com.commercetools.sync.taxcategories.TaxCategorySyncOptions} and a {@link
```
Comment (lojzatran) [inline src/main/java/com/commercetools/sync/sdk2/taxcategories/TaxCategorySync.java:50]:
```suggestion
   * Takes a {@link com.commercetools.sync.sdk2.taxcategories.TaxCategorySyncOptions} and a {@link
```

File: src/main/java/com/commercetools/sync/sdk2/taxcategories/TaxCategorySync.java [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+   * TaxCategorySync} instances to instantiate a new {@link TaxCategorySync} instance that could be
+   * used to sync tax category drafts in the CTP project specified in the injected {@link
+   * com.commercetools.sync.taxcategories.TaxCategorySyncOptions} instance.
```
Comment (lojzatran) [inline src/main/java/com/commercetools/sync/sdk2/taxcategories/TaxCategorySync.java:53]:
```suggestion
   * com.commercetools.sync.sdk2.taxcategories.TaxCategorySyncOptions} instance.
```

File: src/main/java/com/commercetools/sync/sdk2/taxcategories/TaxCategorySync.java [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+      @Nonnull final TaxCategoryService taxCategoryService) {
+    super(new TaxCategorySyncStatistics(), taxCategorySyncOptions);
+    this.taxCategoryService = taxCategoryService;
+    this.batchValidator = new TaxCategoryBatchValidator(getSyncOptions(), getStatistics());
+  }
+
+  @Override
+  protected CompletionStage<TaxCategorySyncStatistics> process(
+      @Nonnull final List<TaxCategoryDraft> resourceDrafts) {
+    List<List<TaxCategoryDraft>> batches =
```
Comment (lojzatran) [inline src/main/java/com/commercetools/sync/sdk2/taxcategories/TaxCategorySync.java:74]:
```suggestion
    final List<List<TaxCategoryDraft>> batches =
```

File: src/main/java/com/commercetools/sync/sdk2/taxcategories/TaxCategorySync.java [hunk 5]
```diff
@@ -0,0 +1,304 @@
...
+      return CompletableFuture.completedFuture(statistics);
+    }
+    final Set<String> validTaxCategoryKeys = result.getRight();
+
+    return taxCategoryService
+        .fetchMatchingTaxCategoriesByKeys(validTaxCategoryKeys)
+        .handle(ImmutablePair::new)
+        .thenCompose(
+            fetchResponse -> {
+              Set<TaxCategory> fetchedTaxCategories = fetchResponse.getKey();
```
Comment (lojzatran) [inline src/main/java/com/commercetools/sync/sdk2/taxcategories/TaxCategorySync.java:114]:
```suggestion
              final Set<TaxCategory> fetchedTaxCategories = fetchResponse.getKey();
```

File: src/main/java/com/commercetools/sync/sdk2/taxcategories/TaxCategorySync.java [hunk 6]
```diff
@@ -0,0 +1,304 @@
...
+  }
+
+  @Nonnull
+  private CompletionStage<Optional<TaxCategory>> buildActionsAndUpdate(
+      @Nonnull final TaxCategory oldTaxCategory, @Nonnull final TaxCategoryDraft newTaxCategory) {
+
+    final List<TaxCategoryUpdateAction> updateActions =
+        buildActions(oldTaxCategory, newTaxCategory);
+
+    List<TaxCategoryUpdateAction> updateActionsAfterCallback =
```
Comment (lojzatran) [inline src/main/java/com/commercetools/sync/sdk2/taxcategories/TaxCategorySync.java:199]:
```suggestion
    final List<TaxCategoryUpdateAction> updateActionsAfterCallback =
```

File: src/main/java/com/commercetools/sync/sdk2/taxcategories/TaxCategorySync.java [hunk 7]
```diff
@@ -0,0 +1,304 @@
...
+      @Nonnull final TaxCategoryDraft newTaxCategory,
+      @Nonnull final List<TaxCategoryUpdateAction> updateActions) {
+
+    return taxCategoryService
+        .updateTaxCategory(oldTaxCategory, updateActions)
+        .handle(ImmutablePair::new)
+        .thenCompose(
+            updateResponse -> {
+              final TaxCategory updatedTaxCategory = updateResponse.getKey();
+              final Throwable sphereException = updateResponse.getValue();
```
Comment (lojzatran) [inline src/main/java/com/commercetools/sync/sdk2/taxcategories/TaxCategorySync.java:236]:
```suggestion
              final Throwable ctpException = updateResponse.getValue();
```

File: src/test/java/com/commercetools/sync/sdk2/taxcategories/TaxCategorySyncOptionsBuilderTest.java [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+import com.commercetools.sync.sdk2.commons.utils.QuadConsumer;
+import com.commercetools.sync.sdk2.commons.utils.TriConsumer;
+import com.commercetools.sync.sdk2.commons.utils.TriFunction;
+import java.util.List;
+import java.util.Optional;
+import org.junit.jupiter.api.Test;
+
+class TaxCategorySyncOptionsBuilderTest {
+
+  private static ProjectApiRoot CTP_CLIENT = mock(ProjectApiRoot.class);
```
Comment (lojzatran) [inline src/test/java/com/commercetools/sync/sdk2/taxcategories/TaxCategorySyncOptionsBuilderTest.java:22]:
```suggestion
  private static final ProjectApiRoot CTP_CLIENT = mock(ProjectApiRoot.class);
```

File: src/test/java/com/commercetools/sync/sdk2/taxcategories/TaxCategorySyncOptionsBuilderTest.java [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+  private TaxCategorySyncOptionsBuilder taxCategorySyncOptionsBuilder =
```
Comment (lojzatran) [inline src/test/java/com/commercetools/sync/sdk2/taxcategories/TaxCategorySyncOptionsBuilderTest.java:23]:
```suggestion
  private final TaxCategorySyncOptionsBuilder taxCategorySyncOptionsBuilder =
```

File: src/test/java/com/commercetools/sync/sdk2/taxcategories/TaxCategorySyncTest.java [hunk 10]
```diff
@@ -0,0 +1,480 @@
...
+        TaxRateDraftBuilder.of()
+            .name(name)
+            .amount(2.0)
+            .includedInPrice(false)
+            .country(CountryCode.FR.getAlpha2())
+            .build();
+    final TaxCategoryDraft draft =
+        TaxCategoryDraftBuilder.of()
+            .name(name)
+            .rates(taxRateFrance, taxRateFrance)
```
Comment (lojzatran) [inline src/test/java/com/commercetools/sync/sdk2/taxcategories/TaxCategorySyncTest.java:442]: Isn't this case similar to the previous one `sync_WithDuplicatedState_ShouldNotBuildActionAndTriggerErrorCallback`?

**Review by lojzatran — CHANGES_REQUESTED** [review]

File: src/test/java/com/commercetools/sync/sdk2/taxcategories/TaxCategorySyncTest.java (hunk 10 above)
Comment (salander85) [reply src/test/java/com/commercetools/sync/sdk2/taxcategories/TaxCategorySyncTest.java:442]: Yes, it is. It's just checkin different error message... this is already checked in batchvalidatortest, so i can remove from here.

**Review by lojzatran — APPROVED** [review]

**Pull request merged**